django_elasticsearch_dsl/signals.py

```diff
@@ -96,6 +99,124 @@ def teardown(self):
         models.signals.m2m_changed.disconnect(self.handle_m2m_changed)
         models.signals.pre_delete.disconnect(self.handle_pre_delete)
 
+try:
+    from celery import shared_task
+except ImportError:
+    pass
+else:
+    class CelerySignalProcessor(RealTimeSignalProcessor):
+        """Celery signal processor.
+
+        Allows automatic updates on the index as delayed background tasks using
+        Celery.
+
+        NB: We cannot process deletes as background tasks.
+        By the time the Celery worker would pick up the delete job, the
+        model instance would already deleted. We can get around this by
+        setting Celery to use `pickle` and sending the object to the worker,
+        but using `pickle` opens the application up to security concerns.
+        """
 
-# Sent after document indexing is completed
-post_index = Signal()
+        def handle_save(self, sender, instance, **kwargs):
+            """Handle save with a Celery task.
+
+            Given an individual model instance, update the object in the index.
+            Update the related objects either.
+            """
+            pk = instance.pk
+            app_label = instance._meta.app_label
+            model_name = instance.__class__.__name__
+
+            self.registry_update_task.delay(pk, app_label, model_name)
+            self.registry_update_related_task.delay(pk, app_label, model_name)
+
+        def handle_pre_delete(self, sender, instance, **kwargs):
+            """Handle removing of instance object from related models instance.
+            We need to do this before the real delete otherwise the relation
+            doesn't exists anymore and we can't get the related models instance.
+            """
+            self.prepare_registry_delete_related_task(instance)
+
+        def handle_delete(self, sender, instance, **kwargs):
+            """Handle delete.
+
+            Given an individual model instance, delete the object from index.
+            """
+            self.prepare_registry_delete_task(instance)
+
+        def prepare_registry_delete_related_task(self, instance):
+            """
+            Select its related instance before this instance was deleted.
+            And pass that to celery.
+            """
+            action = 'index'
+            for doc in registry._get_related_doc(instance):
+                doc_instance = doc(related_instance_to_ignore=instance)
+                try:
+                    related = doc_instance.get_instances_from_related(instance)
+                except ObjectDoesNotExist:
+                    related = None
+                if related is not None:
+                    doc_instance.update(related)
+                    if isinstance(related, models.Model):
+                        object_list = [related]
+                    else:
+                        object_list = related
+                    bulk_data = list(doc_instance._get_actions(object_list, action)),
```

Comment:
`_get_actions` is a private method. I am not sure, whether we should call this private method. Can you come up with any better idea?

Reply:
The public method corresponding to this method has been added